### ISSUE TITLE
question: how to generate globalTypes (like apollo)

### ISSUE THREAD

**Issue by smmoosavi** [relay-tools/relay-compiler-language-typescript#175]
we can generate global types (like enums and custom scalars) with [apollo tooling][apollo-tooling]. you can see `--globalTypesFile` option.

can we do the same thing with relay compiler and generate global types file.

tnx

[apollo-tooling]: https://github.com/apollographql/apollo-tooling

**Comment by renanmav** [comment]
I’m not sure. Maybe @alloy can answer you.

**Comment by sibelius** [comment]
can you provide some example of this?

is this like enums https://github.com/relay-tools/relay-compiler-language-typescript/issues/164?

or custom scalar https://github.com/relay-tools/relay-compiler-language-typescript/issues/129?

**Comment by smmoosavi** [comment]
> or custom scalar #129?

custom scalar is ok.

> is this like enums #164?

this enum type is generated for `Label_label` fragment. but I need some global file that contains all graphql enum types

for example, apollo tooling is generating a file that contained all enum types.

schema:
```graphql
enum Color {
  BLUE
  GREEN
  RED
}
enum Direction { 
  Right
  Left
}
```

```bash
apollo client:codegen --globalTypesFile=src/globalTypes.ts
```

the generated file is independent of any query or fragment in the code and only depends on the schema file.

generated file`src/globalTypes.ts`:
```ts
export enum GenerationType {
  BLUE = "BLUE",
  GREEN = "GREEN",
  RED = "RED",
}
export enum Direction {/*..*/}
// and all other enums defined in the schema file
```
for the relay, it would be like `src/globalTypes.ts`:
```ts
export type Color = "BLUE" | "GREEN" | "RED" | "%future added value";
export type Direction = "..."
```

**Comment by smmoosavi** [comment]
for example, the `ColorSelector` component doesn't have any fragment or query. but it needs `Color` enum. and shouldn't depend on any of the other usages that have the enum definitions.

**Comment by sibelius** [comment]
you want something like this

https://github.com/facebook/relay/blob/master/packages/relay-compiler/core/RelayGraphQLEnumsGenerator.js

`RelayGraphQLEnumsGenerator` is only for `Flow` right now

can you try implement this to Typescript?

**Comment by smmoosavi** [comment]
> you want something like this

yes. it is what I want (I assume [schema argument][schema] is all defined types in graphql)

> can you try to implement this to Typescript?

yes, I can. flow and typescript output is almost equal in this case. except for the header part.

but I don't know at which file it should be written. or where and how this should be used.


[schema]: https://github.com/facebook/relay/blob/57c45c2c3bafb2dec9a32a824e4e8bc4aef4a402/packages/relay-compiler/core/RelayGraphQLEnumsGenerator.js#L21

**Comment by c0m1t** [comment]
https://github.com/facebook/relay/issues/3051

**Comment by c0m1t** [comment]
Me and @smmoosavi wrote a small package to solve this issue: `relay-enum-generator`
[Link](https://github.com/c0m1t/relay-enum-generator)

**Comment by sibelius** [comment]
Great to hear, can you add this to relay documentation?

**Comment by c0m1t** [comment]
@sibelius  Where should I add this to?

**Comment by sibelius** [comment]
Check Facebook/relay repo, there is a docusaurus there, this could be close to type emission guide

**Comment by DamodarSojka** [comment]
How about just type aliases?
for example:
export type UUID = string;
The type is of course not imported by all the __generated__ files making them fail the compilation.
